### PR TITLE
Pull request - extended circulant functionality + DFT representation.

### PR DESCRIPTION
**Pull request by willtebbutt**
Hi.

I have extended the package in a couple of ways. Firstly, the Discrete Fourier Transform (DFT) matrix has been added in src/DFT.jl. This is interesting semantically as it allows for (in certain circumstances) a more natural way of expressing the Fourier Transform of vectors and matrices (where a matrix is just viewed as a collection of vectors to be transformed). Note that there are primarily two implementations of this matrix, Unitary DFT (UDFT) uses the matrix U defined such that ctranspose(U) = inv(U), whereas this does not hold for DFT.

This is particularly useful for cleaning up the implementation of the Circulant matrix, as we can now describe all of the operations one might expect to be able to perform with Circulant matrices in terms of multiplication by the DFT matrix or it's conjugate transpose (or some re-scaling thereof). Furthermore, I have added a significant amount of extra functionality involving Circulant matrices, the vast majority of which is tested in test/circulant.jl. (Note that I have separated out the functionality involving Toeplitz and Circulant matrices). Furthermore, I have changed the behaviour of the multiplication of real and integer Circulant matrices. Multiplication now always returns a complex-valued float, but helper functions have been added (see the end of src/circulant.jl) to extend the casting functionality from Base necessary to enable the user to explicitly round and cast. It is my feeling that this is the correct thing to do as it is unwise to mask type conversions that may result in some loss of accuracy from the user. Commenting in src/circulant.jl is fairly sparse as most code should be self-documenting. Let me know if more comments are wanted.

I appreciate that this is a fairly substantial modification / addition, so please review and let me know what you think!

### CODE REVIEW

**Comment by ivanslapnicar** [comment]
Hi,


there is lot of changes which are not so easy to follow.

The previous approach with just fft and ifft needs O(n) operations.

This is especially interesting when using Lanczos method to compute SOME eigenvalues and eigenvectors (or singular values and vectors)

of Hankel and Toplitz matrices using eigs() (svds() ), since the multiplication can be defined as an operator.

For large Hankel matrix (say n=50.000) computation of 40 eigenpairs with this approach is very fast (5 seconds).

On the the other hand, the full eigenvalue matrix of the underlying Circulant cannot be stored at all.

So, I just want to make sure that the new approach is the same in terms of operation count.


Cheers, Ivan


Prof. dr. sc. Ivan Slapničar
Sveučilište u Splitu / University of Split
FESB
R. Boškovića 32
HR-21000 Split
Hrvatska / Croatia
Phone: + 385 21 305893
Ured/Office: B803


________________________________
Šalje: Will Tebbutt <notifications@github.com>
Poslano: 19. siječnja 2017. 19:10
Prima: jiahao/SpecialMatrices.jl
Kopija: Subscribed
Predmet: [jiahao/SpecialMatrices.jl] Pull request - extended circulant functionality + DFT representation. (#13)


Hi.

I have extended the package in a couple of ways. Firstly, the Discrete Fourier Transform (DFT) matrix has been added in src/DFT.jl. This is interesting semantically as it allows for (in certain circumstances) a more natural way of expressing the Fourier Transform of vectors and matrices (where a matrix is just viewed as a collection of vectors to be transformed). Note that there are primarily two implementations of this matrix, Unitary DFT (UDFT) uses the matrix U defined such that ctranspose(U) = inv(U), whereas this does not hold for DFT.

This is particularly useful for cleaning up the implementation of the Circulant matrix, as we can now describe all of the operations one might expect to be able to perform with Circulant matrices in terms of multiplication by the DFT matrix or it's conjugate transpose (or some re-scaling thereof). Furthermore, I have added a significant amount of extra functionality involving Circulant matrices, the vast majority of which is tested in test/circulant.jl. (Note that I have separated out the functionality involving Toeplitz and Circulant matrices). Furthermore, I have changed the behaviour of the multiplication of real and integer Circulant matrices. Multiplication now always returns a complex-valued float, but helper functions have been added (see the end of src/circulant.jl) to extend the casting functionality from Base necessary to enable the user to explicitly round and cast. It is my feeling that this is the correct thing to do as it is unwise to mask type conversions that may result in some loss of accuracy from the user. Commenting in src/circulant.jl is fairly sparse as most code should be self-documenting. Let me know if more comments are wanted.

I appreciate that this is a fairly substantial modification / addition, so please review and let me know what you think!

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/jiahao/SpecialMatrices.jl/pull/13

[https://avatars0.githubusercontent.com/u/3628294?v=3&s=400]<https://github.com/jiahao/SpecialMatrices.jl/pull/13>

Pull request - extended circulant functionality + DFT representation. by willtebbutt · Pull Request #13 · jiahao/SpecialMatrices.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13>
github.com
Hi. I have extended the package in a couple of ways. Firstly, the Discrete Fourier Transform (DFT) matrix has been added in src/DFT.jl. This is interesting semantically as it allows for (in certain...

Commit Summary

  *   Basic DFT functionality added. Coverage of the expected operations is, however, somewhat patchy.
  *   Re-implmented existing Circulant multiplication functionality to correctly use dispatch to determine what to cast back to. Previous implementation used dynamic type checking and missed some edge cases. Not re-implemented the in-place multiplication functionality in this commit, nor have the tests been updated. Also added typealias for Circulant to reduce verbosity.
  *   Completed + tested basic functionality for circulant matrices, including multiplication in terms of DFT matrices, (conjugate) transposition, addition, subtraction, eigenvalues + eigenvectors. dft.jl has been further extended to cover more functionality and a few bugs have been fixed.
  *   Commit to bring remote up to speed.
  *   Circulant functionality sufficiently complete for a first vesion.

File Changes

  *   M src/SpecialMatrices.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-0> (32)
  *   A src/circulant.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-1> (124)
  *   A src/dft.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-2> (86)
  *   M src/hankel.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-3> (3)
  *   M src/toeplitz.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-4> (47)
  *   A test/circulant.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-5> (127)
  *   A test/dft.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-6> (19)
  *   M test/runtests.jl<https://github.com/jiahao/SpecialMatrices.jl/pull/13/files#diff-7> (6)

Patch Links:

  *   https://github.com/jiahao/SpecialMatrices.jl/pull/13.patch
  *   https://github.com/jiahao/SpecialMatrices.jl/pull/13.diff

**Comment by willtebbutt** [comment]
Hi Ivan,

Thanks for the response. With regards to your concerns about the underlying Circulant implementation, I have been careful to ensure that (U)DFT matrices are never explicitly evaluated - as you point out doing so would require `O(n^2)` operations + memory which would indeed defeat the point of working with Circulant matrices to get the complexity down to `O(n logn)`.

As evidence for this, I would first highlight lines 10-14 of `src/dft.jl`. The UDFT (Unitary DFT, eigenvectors of the Circulant matrix of a given size) stores only a single Int and Float64 (the latter being for convenience). Furthermore, there is no indexing or 'full' functionality defined for the (U)DFT matrix, so it is not possible for the JIT to unexpectedly compile code which explicitly evaluates the full (U)DFT matrix and default to a naive approach to performing multiplication operations. Furthermore, if you look at, for example, line 45 of `src/dft.jl` you will see that the left-multiplication of a StridedVecOrMat (SVM) by the DFT matrix is implemented in terms of the fft, as one would hope. The rest of the file is more of the same - it is all convenience functionality to mean that you can write things like
~~~~
X * U
~~~~
for some StridedVecOrMat X and (U)DFT U, and the correct (i)fft operation will be used to compute the requested operations efficiently under the hood.

This propagates through to the Circulant matrices in `src/circulant.jl`. First observe line 41; this line 'computes' the eigen-factorisation of a circulant matrix without actually computing the eigenvectors explicitly, by using the previously described UDFT matrix. This means that multiplication routine on line 64 works efficiently. I'll explain what is going on there: `conf(U,X)` checks that `U` and `X` are conformal. Line 66 simply gets pointers to the eigenvalues and stores them in a `Diagonal` matrix (which is implemented efficiently under the hood to do the things that you would expect), and get a pointer to the UDFT matrix which represents the eigenvectors. Consequently, line 67 is computed efficiently using the multiplication function on line 47 of `src/dft.jl`. `Γ * Y` on line 68 is computed efficiently by exploiting the efficient representation of `Γ` and the `Ac_mul_B!` is computed efficiently using code from line 77 of `src/dft.jl`.

Note that the discussed multiplication function accepts the eigen-factorisation of some Circulant matrix. This is done so that if one is, for example, left-multiplying multiple vectors/matrices by a Circulant matrix, the initial fft required to obtain the eigenvalues can be done once and cached (this is relevant to some of the work that I am doing). However, line 76 of `src/circulant.jl` provides a wrapper which computes the eigen-factorisation and passes it on such that a 'raw' Circulant matrix can be provided if the described caching behaviour is not required.

The rest of the functionality in `src/circulant.jl` is more of the same - care has been taken to ensure that all operations have at most complexity `O(n logn)`.

The only notable changes to your API are that the indexing on line 18 of `src/circulant.jl` has been commented out - depending upon your use case it might make sense for this not to be the case.

Best,
Will

File: src/circulant.jl [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+-(a::Number, C::Circ) = Circ(a - C.c)
+-(C::Circ, D::Circ) = Circ(C.c - D.c)
+-(C::Circ) = Circ(-C.c)
+
+# Return the eigen-factorisation of C. Constituents are efficient to compute
+# and the eigenvectors are represented efficiently using a UnitaryDFT object.
+eig{T<:Real}(C::Circ{T}) = (N = length(C.c); (DFT{Complex{T}}(N) * C.c, UDFT{Complex{T}}(N)))
+eig{T<:Complex}(C::Circ{T}) = (N = length(C.c); (DFT{T}(N) * C.c, UDFT{T}(N)))
+eigfact(C::Circ) = Eigen(eig(C)...)
+typealias CircEig{T<:Complex} Eigen{T, T, UDFT{T}, Vector{T}}
```
Comment (iamed2) [inline src/circulant.jl]: Calling `show` on an instance of this type causes a StackOverflowError.

`show(STDOUT, eigfact(Circulant(collect(1.0:10.0))));` to replicate

File: src/circulant.jl [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+-(C::Circ, a::Number) = Circ(C.c - a)
+-(a::Number, C::Circ) = Circ(a - C.c)
+-(C::Circ, D::Circ) = Circ(C.c - D.c)
+-(C::Circ) = Circ(-C.c)
+
+# Return the eigen-factorisation of C. Constituents are efficient to compute
+# and the eigenvectors are represented efficiently using a UnitaryDFT object.
+eig{T<:Real}(C::Circ{T}) = (N = length(C.c); (DFT{Complex{T}}(N) * C.c, UDFT{Complex{T}}(N)))
+eig{T<:Complex}(C::Circ{T}) = (N = length(C.c); (DFT{T}(N) * C.c, UDFT{T}(N)))
+eigfact(C::Circ) = Eigen(eig(C)...)
```
Comment (iamed2) [inline src/circulant.jl]: This function should convert/promote to the right types such that the answer is a `CircEig`. Currently that is not the case for `Circ{T} where T <: Integer`. That should either be fixed or disallowed.
